Display departments

heatlikeheatwave opened the pull request:
## 🗒️ What

- Updated types for `Person` to support required data to build out feature (`title`, `responsiveImage` properties `srcSet` and `sizes`)
- Updated types for `Department` to support `parent` and `children` properties
- Added a `hierarchyBuilder` function to take in a list with `parent` and `children` properties
- Added a `DepartmentList` component to display the departments

## 🤷 Why

- `DepartmentList` component support arbitrary nesting, regardless of data structure nesting

<img width="712" alt="Screenshot 2023-08-25 at 16 15 30" src="https://github.com/heatlikeheatwave/hc-webdev-challenge/assets/55773810/aa01159d-a18a-4a49-a7ce-e2a74daa50cd">

## 🛠️ How

My first approach to support arbitrary nesting was to modify the data returned by dictating the levels of `children` properties the query should return past the current data (e.g. return 6 levels of children, even though the current department list only requires 3). I also looked into recursively fetching the children data as referenced (here)[https://github.com/graphql/graphql-spec/issues/91#issuecomment-197860064].

I decided against those approaches after reading this issue thread in the [graphql spec](https://github.com/graphql/graphql-spec/issues/91). Building the department list hierarchy from the returned data's `children` and `parent` properties seemed like the most flexible way to support nesting without requiring the query to be changed in the future, while limiting development time. 

## 📸 Design Screenshots
<img width="424" alt="Screenshot 2023-08-25 at 16 32 45" src="https://github.com/heatlikeheatwave/hc-webdev-challenge/assets/55773810/a554dc50-78af-4d28-9c04-daeeca3b9b79">

## 💭 Anything else?
I'm splitting up the display of people data and search functionality into other PRs to ensure each one doesn't take up too much time for the reviewer(s) and is easy to review

dstaley commented on helpers/buildHierarchy.ts:
```diff
@@ -0,0 +1,20 @@
+			}
+		}
+		return acc
+	}, [])
```
```suggestion
	}, [] as Department[])
```

Adding a case ensures that your `acc` argument gets type checked.

dstaley reviewed (commented):
This is a great start! I really appreciate you breaking this down into easier-to-digest pull requests. Getting the department tree data structure is definitely tricky, so I'm glad to see you tackling that first. I think the biggest change I'd like to see is a `buildHierarchy` function that works with arbitrarily sorted departments, since there's no guarantee that the CMS will always return sorted data (currently I believe the results are sorted by last updated first). The other suggestions are mostly TypeScript improvements that will help make your code more type-safe.

heatlikeheatwave commented:
> This is a great start! I really appreciate you breaking this down into easier-to-digest pull requests. Getting the department tree data structure is definitely tricky, so I'm glad to see you tackling that first. I think the biggest change I'd like to see is a `buildHierarchy` function that works with arbitrarily sorted departments, since there's no guarantee that the CMS will always return sorted data (currently I believe the results are sorted by last updated first). The other suggestions are mostly TypeScript improvements that will help make your code more type-safe.

Thank you for the feedback! I changed the `buildHierarchy` function in [this commit](https://github.com/heatlikeheatwave/hc-webdev-challenge/pull/2/commits/7a978f8cf79334839971c746e7dc72d584e7c13b) to nest a flat array based on parent id. I also added the type checking you recommended. Would love your thoughts on it now!

heatlikeheatwave commented on helpers/buildHierarchy.ts:
```diff
@@ -1,20 +1,27 @@
 import { Department } from 'types'
+type DepartmentRecord = Record<string, Department>
```
Removed in another commit

dstaley commented on helpers/buildHierarchy.ts:
```diff
@@ -1,20 +1,30 @@
+function buildChildren(
+	list: DepartmentRecord[]
+): Record<string, DepartmentRecord> {
+	return list.reduce((list, item) => {
```
For clarity, we should use a different name for the accumulator variable so as not to confuse it with the `list` parameter.

dstaley commented on helpers/buildHierarchy.ts:
```diff
@@ -1,20 +1,30 @@
+	}, {} as Record<string, DepartmentRecord>)
+}
+
+export function buildHierarchy(
```
This looks great! I think it's _just_ over the threshold to necessitate comments explaining what happens. I tested with various orderings and it looks like it works correctly regardless of order, so great job!
